lib/runtime/src/ops/concat.cc

```diff
@@ -16,537 +16,199 @@
 
-Concat::Concat(FFModel &model,
-               int _n,
-               ParallelTensor const *_tensors,
-               int _legion_axis,
-               char const *name)
-    : Op(model,
-         OP_CONCAT,
-         DT_FLOAT,
-         name,
-         _n /*inputs*/,
-         0 /*weights*/,
-         1 /*outputs*/,
-         _tensors),
-      legion_axis(_legion_axis) {
-  int num_dim = inputs[0]->num_dims;
-  ParallelDim dims[MAX_TENSOR_DIM];
-  for (int i = 0; i < num_dim; i++) {
-    dims[i] = inputs[0]->dims[i];
-  }
-  for (int i = 1; i < numInputs; i++) {
-    assert(inputs[i]->data_type == inputs[0]->data_type);
-    assert(inputs[i]->num_dims == inputs[0]->num_dims);
-    for (int j = 0; j < num_dim; j++) {
-      if (j != legion_axis) {
-        assert(inputs[i]->dims[j] == inputs[0]->dims[j]);
-      } else {
-        // Assert that the concat dim cannot be parallelized
-        assert(inputs[i]->dims[j].parallel_idx == -1);
-        assert(inputs[i]->dims[j].degree == 1);
-        dims[j].size += inputs[i]->dims[j].size;
-      }
-    }
-  }
-  numOutputs = 1;
-  outputs[0] = model.create_parallel_tensor_legion_ordering(
-      num_dim, dims, inputs[0]->data_type, this);
+  return {CONCAT_FWD_TASK_ID, binding};
 }
 
-Concat::Concat(FFModel &model,
-               ConcatParams const &params,
-               std::vector<ParallelTensor> const &inputs,
-               char const *name)
-    : Concat(model, inputs.size(), inputs.data(), params.axis, name) {}
+OpTaskInvocation backward(ConcatAttrs const &attrs) {
+  OpTaskBinding b = infer_bwd_binding(forward(attrs).binding);
 
-static OpTaskSignature get_init_task_signature() {
-  OpTaskSignature init(OpTaskType::INIT);
+  return {CONCAT_BWD_TASK_ID, b};
+}
 
-  init.add_arg_slot<ConcatAttrs>(ATTRS);
-  init.add_arg_slot<bool>(PROFILING);
+static DeviceSpecific<ConcatPerDeviceState>
+    init_task_impl(TaskArgumentAccessor const &acc) {
+  auto const &attrs = acc.get_argument<ConcatAttrs>(ATTRS);
+  PerDeviceFFHandle handle = acc.get_argument<PerDeviceFFHandle>(HANDLE);
 
-  init.add_input_slot(INPUTS, SlotType::VARIADIC);
-  init.add_output_slot(OUTPUT);
+  DeviceSpecific<ConcatPerDeviceState> per_device_state =
+      acc.create_device_specific<ConcatPerDeviceState>(init_kernel(attrs.axis));
+  return per_device_state;
+}
 
-  return init;
+static DeviceSpecific<ConcatPerDeviceState>
+    init_task(Task const *task,
+              std::vector<PhysicalRegion> const &regions,
+              Context ctx,
+              Runtime *runtime) {
+  TaskArgumentAccessor acc(task, regions, ctx, runtime);
+  return init_task_impl(acc);
 }
 
-static OpTaskSignature get_fwd_task_signature() {
-  OpTaskSignature fwd(OpTaskType::FWD);
+static optional<float> forward_task_impl(TaskArgumentAccessor const &acc) {
+  auto per_device_state =
+      acc.get_argument<ConcatPerDeviceState>(PER_DEVICE_STATE);
+  int number_of_inputs = acc.get_argument<int>(NUM_INPUTS);
```

Comment:
Should just have `n` be in `ConcatAttrs`. Then you can `acc.get_argument<ConcatAttrs>(ATTRS)` and then get number of inputs from there.